### PR TITLE
Migrate S2259 (NullDereferenceCheck) to eslint implementation

### LINKED ISSUE #2331: Migrate S2259 (NullDereferenceCheck) to eslint implementation

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #2331

### CODE REVIEW

File: eslint-bridge/tests/rules/null-dereference.test.ts [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+              x.call();
+            }
+          
+            var y = foo();
+          
+            if (typeof y === 'undefined') {
+              y.call();  // Noncompliant
+            }
+          }`,
+      errors: 1,
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/tests/rules/null-dereference.test.ts:360]: Can you maybe add an assertion on the location for this issue?

File: eslint-bridge/tests/rules/null-dereference.test.ts [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+      
+        var initializedArray = [];
+        for(let i of initializedArray) {      // OK
+        }
+      
+        var x;
+        for(x of obj) {                       // OK we should not care about x being undefined
+        }
+      }`,
+      errors: 2,
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/tests/rules/null-dereference.test.ts:392]: Same here, I feel it's better to be 100% sure the issue is at the right place.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM.
Although I think that, while we lose quite a few FPs from the change of implementation, we also lose interesting TPs as well. Do you think it'd be worth documenting somehow the rules for which it happens for this last batch of migration? 
I feel it's  bound to happen for some of the remaining rules and while it shouldn't stop us from finishing the migration, it could be interesting to keep a list of rules that we know we can improve later on.

**Comment by andrea-guarino-sonarsource** [comment]
As discussed, I will create tickets to document those introduced FN

**Pull request merged**